rust/main/submitter/src/payload_dispatcher/stages/inclusion_stage.rs

```diff
@@ -1 +1,406 @@
+// TODO: re-enable clippy warnings
+#![allow(dead_code)]
 
+use std::{
+    collections::{HashMap, VecDeque},
+    future::Future,
+    sync::Arc,
+    time::Duration,
+};
+
+use derive_new::new;
+use eyre::{eyre, Result};
+use futures_util::future::try_join_all;
+use tokio::{
+    sync::{mpsc, Mutex},
+    time::sleep,
+};
+use tracing::{error, info, info_span, Instrument};
+
+use crate::{
+    payload::{FullPayload, PayloadStatus},
+    transaction::{DropReason as TxDropReason, Transaction, TransactionId, TransactionStatus},
+};
+
+use super::{utils::retry_until_success, PayloadDispatcherState};
+
+pub type InclusionStagePool = Arc<Mutex<HashMap<TransactionId, Transaction>>>;
+
+#[derive(new)]
+struct InclusionStage {
+    pool: InclusionStagePool,
+    building_stage_receiver: mpsc::Receiver<Transaction>,
+    finality_stage_sender: mpsc::Sender<Transaction>,
+    state: PayloadDispatcherState,
+}
+
+impl InclusionStage {
+    pub async fn run(self) {
+        let InclusionStage {
+            pool,
+            building_stage_receiver,
+            finality_stage_sender,
+            state,
+        } = self;
+        let futures = vec![
+            tokio::spawn(
+                Self::receive_txs(building_stage_receiver, pool.clone())
+                    .instrument(info_span!("receive_txs")),
+            ),
+            tokio::spawn(
+                Self::process_txs(pool, finality_stage_sender, state)
+                    .instrument(info_span!("process_txs")),
+            ),
+        ];
+        if let Err(err) = try_join_all(futures).await {
+            error!(
+                error=?err,
+                "Inclusion stage future panicked"
+            );
+        }
+    }
+
+    async fn receive_txs(
+        mut building_stage_receiver: mpsc::Receiver<Transaction>,
+        pool: InclusionStagePool,
+    ) -> Result<()> {
+        loop {
+            let tx = building_stage_receiver.recv().await.unwrap();
```

Comment:
If `recv` returns `None`, the channel is closed. I suppose we can raise a critical error if it happens so that we don't crash the whole relayer.

Reply:
ah my bad, this `unwrap` wasn't supposed to make its way here